diff --git a/flink-core/src/main/java/org/apache/flink/configuration/ConfigOption.java b/flink-core/src/main/java/org/apache/flink/configuration/ConfigOption.java
index 9ad49c09..10cdc3f1 100644
--- a/flink-core/src/main/java/org/apache/flink/configuration/ConfigOption.java
+++ b/flink-core/src/main/java/org/apache/flink/configuration/ConfigOption.java
@@ -43,6 +43,8 @@ public class ConfigOption<T> {
 
 	private static final FallbackKey[] EMPTY = new FallbackKey[0];
 
+	static final Description EMPTY_DESCRIPTION = Description.builder().text("").build();
+
 	// ------------------------------------------------------------------------
 
 	/** The current key for that config option. */
@@ -57,51 +59,51 @@ public class ConfigOption<T> {
 	/** The description for this option. */
 	private final Description description;
 
-	// ------------------------------------------------------------------------
-
 	/**
-	 * Creates a new config option with no fallback keys.
-	 *
-	 * @param key The current key for that config option
-	 * @param defaultValue The default value for this option
+	 * Type of the value that this ConfigOption describes.
+	 * <ul>
+	 *     <li>typeClass == atomic class (e.g. {@code Integer.class}) -> {@code ConfigOption<Integer>}</li>
+	 *     <li>typeClass == {@code Map.class} -> {@code ConfigOption<Map<String, String>>}</li>
+	 *     <li>typeClass == atomic class and isList == true for {@code ConfigOption<List<Integer>>}</li>
+	 * </ul>
 	 */
-	ConfigOption(String key, T defaultValue) {
-		this.key = checkNotNull(key);
-		this.description = Description.builder().text("").build();
-		this.defaultValue = defaultValue;
-		this.fallbackKeys = EMPTY;
+	private final Class clazz;
+
+	private final boolean isList;
+
+	// ------------------------------------------------------------------------
+
+	Class getClazz() {
+		return clazz;
 	}
 
-	/**
-	 * Creates a new config option with fallback keys.
-	 *
-	 * @param key The current key for that config option
-	 * @param description Description for that option
-	 * @param defaultValue The default value for this option
-	 * @param fallbackKeys The list of fallback keys, in the order to be checked
-	 * @deprecated use version with {@link Description} instead
-	 */
-	@Deprecated
-	ConfigOption(String key, String description, T defaultValue, FallbackKey... fallbackKeys) {
-		this.key = checkNotNull(key);
-		this.description = Description.builder().text(description).build();
-		this.defaultValue = defaultValue;
-		this.fallbackKeys = fallbackKeys == null || fallbackKeys.length == 0 ? EMPTY : fallbackKeys;
+	boolean isList() {
+		return isList;
 	}
 
 	/**
 	 * Creates a new config option with fallback keys.
 	 *
 	 * @param key The current key for that config option
+	 * @param clazz describes type of the ConfigOption, see description of the clazz field
 	 * @param description Description for that option
 	 * @param defaultValue The default value for this option
+	 * @param isList tells if the ConfigOption describes a list option, see description of the clazz field
 	 * @param fallbackKeys The list of fallback keys, in the order to be checked
 	 */
-	ConfigOption(String key, Description description, T defaultValue, FallbackKey... fallbackKeys) {
+	ConfigOption(
+			String key,
+			Class clazz,
+			Description description,
+			T defaultValue,
+			boolean isList,
+			FallbackKey... fallbackKeys) {
 		this.key = checkNotNull(key);
 		this.description = description;
 		this.defaultValue = defaultValue;
 		this.fallbackKeys = fallbackKeys == null || fallbackKeys.length == 0 ? EMPTY : fallbackKeys;
+		this.clazz = checkNotNull(clazz);
+		this.isList = isList;
 	}
 
 	// ------------------------------------------------------------------------
@@ -124,7 +126,7 @@ public class ConfigOption<T> {
 		// put fallback keys first so that they are prioritized
 		final FallbackKey[] mergedAlternativeKeys = Stream.concat(newFallbackKeys, currentAlternativeKeys)
 			.toArray(FallbackKey[]::new);
-		return new ConfigOption<>(key, description, defaultValue, mergedAlternativeKeys);
+		return new ConfigOption<>(key, clazz, description, defaultValue, isList, mergedAlternativeKeys);
 	}
 
 	/**
@@ -145,7 +147,7 @@ public class ConfigOption<T> {
 		// put deprecated keys last so that they are de-prioritized
 		final FallbackKey[] mergedAlternativeKeys = Stream.concat(currentAlternativeKeys, newDeprecatedKeys)
 			.toArray(FallbackKey[]::new);
-		return new ConfigOption<>(key, description, defaultValue, mergedAlternativeKeys);
+		return new ConfigOption<>(key, clazz, description, defaultValue, isList, mergedAlternativeKeys);
 	}
 
 	/**
@@ -167,7 +169,7 @@ public class ConfigOption<T> {
 	 * @return A new config option, with given description.
 	 */
 	public ConfigOption<T> withDescription(final Description description) {
-		return new ConfigOption<>(key, description, defaultValue, fallbackKeys);
+		return new ConfigOption<>(key, clazz, description, defaultValue, isList, fallbackKeys);
 	}
 
 	// ------------------------------------------------------------------------
@@ -203,8 +205,7 @@ public class ConfigOption<T> {
 	 */
 	@Deprecated
 	public boolean hasDeprecatedKeys() {
-		return fallbackKeys == EMPTY ? false :
-			Arrays.stream(fallbackKeys).anyMatch(FallbackKey::isDeprecated);
+		return fallbackKeys != EMPTY && Arrays.stream(fallbackKeys).anyMatch(FallbackKey::isDeprecated);
 	}
 
 	/**
@@ -276,4 +277,5 @@ public class ConfigOption<T> {
 		return String.format("Key: '%s' , default: %s (fallback keys: %s)",
 				key, defaultValue, Arrays.toString(fallbackKeys));
 	}
+
 }
diff --git a/flink-core/src/main/java/org/apache/flink/configuration/ConfigOptions.java b/flink-core/src/main/java/org/apache/flink/configuration/ConfigOptions.java
index 1ec5b3ca..559cbf46 100644
--- a/flink-core/src/main/java/org/apache/flink/configuration/ConfigOptions.java
+++ b/flink-core/src/main/java/org/apache/flink/configuration/ConfigOptions.java
@@ -19,6 +19,12 @@
 package org.apache.flink.configuration;
 
 import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.configuration.description.Description;
+
+import java.time.Duration;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
 
 import static org.apache.flink.util.Preconditions.checkNotNull;
 
@@ -30,21 +36,32 @@ import static org.apache.flink.util.Preconditions.checkNotNull;
  * // simple string-valued option with a default value
  * ConfigOption<String> tempDirs = ConfigOptions
  *     .key("tmp.dir")
+ *     .stringType()
  *     .defaultValue("/tmp");
  *
  * // simple integer-valued option with a default value
  * ConfigOption<Integer> parallelism = ConfigOptions
  *     .key("application.parallelism")
+ *     .intType()
  *     .defaultValue(100);
  *
+ * // option of list of integers with a default value
+ * ConfigOption<Integer> parallelism = ConfigOptions
+ *     .key("application.ports")
+ *     .intType()
+ *     .asList()
+ *     .defaultValue(8000, 8001, 8002);
+ *
  * // option with no default value
  * ConfigOption<String> userName = ConfigOptions
  *     .key("user.name")
+ *     .stringType()
  *     .noDefaultValue();
  *
  * // option with deprecated keys to check
  * ConfigOption<Double> threshold = ConfigOptions
  *     .key("cpu.utilization.threshold")
+ *     .doubleType()
  *     .defaultValue(0.9).
  *     .withDeprecatedKeys("cpu.threshold");
  * }</pre>
@@ -82,6 +99,79 @@ public class ConfigOptions {
 			this.key = key;
 		}
 
+		/**
+		 * Defines that the value of the option should be of {@link Boolean} type.
+		 */
+		public TypedConfigOptionBuilder<Boolean> booleanType() {
+			return new TypedConfigOptionBuilder<>(key, Boolean.class);
+		}
+
+		/**
+		 * Defines that the value of the option should be of {@link Integer} type.
+		 */
+		public TypedConfigOptionBuilder<Integer> intType() {
+			return new TypedConfigOptionBuilder<>(key, Integer.class);
+		}
+
+		/**
+		 * Defines that the value of the option should be of {@link Long} type.
+		 */
+		public TypedConfigOptionBuilder<Long> longType() {
+			return new TypedConfigOptionBuilder<>(key, Long.class);
+		}
+
+		/**
+		 * Defines that the value of the option should be of {@link Float} type.
+		 */
+		public TypedConfigOptionBuilder<Float> floatType() {
+			return new TypedConfigOptionBuilder<>(key, Float.class);
+		}
+
+		/**
+		 * Defines that the value of the option should be of {@link Double} type.
+		 */
+		public TypedConfigOptionBuilder<Double> doubleType() {
+			return new TypedConfigOptionBuilder<>(key, Double.class);
+		}
+
+		/**
+		 * Defines that the value of the option should be of {@link String} type.
+		 */
+		public TypedConfigOptionBuilder<String> stringType() {
+			return new TypedConfigOptionBuilder<>(key, String.class);
+		}
+
+		/**
+		 * Defines that the value of the option should be of {@link Duration} type.
+		 */
+		public TypedConfigOptionBuilder<Duration> durationType() {
+			return new TypedConfigOptionBuilder<>(key, Duration.class);
+		}
+
+		/**
+		 * Defines that the value of the option should be of {@link MemorySize} type.
+		 */
+		public TypedConfigOptionBuilder<MemorySize> memoryType() {
+			return new TypedConfigOptionBuilder<>(key, MemorySize.class);
+		}
+
+		/**
+		 * Defines that the value of the option should be of {@link Enum} type.
+		 *
+		 * @param enumClass Concrete type of the expected enum.
+		 */
+		public <T extends Enum<T>> TypedConfigOptionBuilder<T> enumType(Class<T> enumClass) {
+			return new TypedConfigOptionBuilder<>(key, enumClass);
+		}
+
+		/**
+		 * Defines that the value of the option should be a set of properties, which can be represented as
+		 * {@code Map<String, String>}.
+		 */
+		public TypedConfigOptionBuilder<Map<String, String>> mapType() {
+			return new TypedConfigOptionBuilder<>(key, Map.class);
+		}
+
 		/**
 		 * Creates a ConfigOption with the given default value.
 		 *
@@ -91,10 +181,17 @@ public class ConfigOptions {
 		 * @param value The default value for the config option
 		 * @param <T> The type of the default value.
 		 * @return The config option with the default value.
+		 * @deprecated define the type explicitly first with one of the intType(), stringType(), etc.
 		 */
+		@Deprecated
 		public <T> ConfigOption<T> defaultValue(T value) {
 			checkNotNull(value);
-			return new ConfigOption<>(key, value);
+			return new ConfigOption<>(
+				key,
+				value.getClass(),
+				ConfigOption.EMPTY_DESCRIPTION,
+				value,
+				false);
 		}
 
 		/**
@@ -103,12 +200,116 @@ public class ConfigOptions {
 		 * default value.
 		 *
 		 * @return The created ConfigOption.
+		 * @deprecated define the type explicitly first with one of the intType(), stringType(), etc.
 		 */
+		@Deprecated
 		public ConfigOption<String> noDefaultValue() {
-			return new ConfigOption<>(key, null);
+			return new ConfigOption<>(
+				key,
+				String.class,
+				ConfigOption.EMPTY_DESCRIPTION,
+				null,
+				false);
+		}
+	}
+
+	/**
+	 * Builder for {@link ConfigOption} with a defined atomic type.
+	 *
+	 * @param <T> atomic type of the option
+	 */
+	public static class TypedConfigOptionBuilder<T> {
+		private final String key;
+		private final Class clazz;
+
+		TypedConfigOptionBuilder(String key, Class clazz) {
+			this.key = key;
+			this.clazz = clazz;
+		}
+
+		/**
+		 * Defines that the option's type should be a list of previously defined atomic type.
+		 */
+		public ListConfigOptionBuilder<T> asList() {
+			return new ListConfigOptionBuilder<>(key, clazz);
+		}
+
+		/**
+		 * Creates a ConfigOption with the given default value.
+		 *
+		 * @param value The default value for the config option
+		 * @return The config option with the default value.
+		 */
+		public ConfigOption<T> defaultValue(T value) {
+			return new ConfigOption<>(
+				key,
+				clazz,
+				ConfigOption.EMPTY_DESCRIPTION,
+				value,
+				false);
+		}
+
+		/**
+		 * Creates a ConfigOption without a default value.
+		 *
+		 * @return The config option without a default value.
+		 */
+		public ConfigOption<T> noDefaultValue() {
+			return new ConfigOption<T>(
+				key,
+				clazz,
+				Description.builder().text("").build(),
+				null,
+				false);
 		}
 	}
 
+	/**
+	 * Builder for {@link ConfigOption} of list of type T.
+	 *
+	 * @param <T> atomic type of the option
+	 */
+	public static class ListConfigOptionBuilder<T> {
+		private final String key;
+		private final Class clazz;
+
+		ListConfigOptionBuilder(String key, Class clazz) {
+			this.key = key;
+			this.clazz = clazz;
+		}
+
+		/**
+		 * Creates a ConfigOption with the given default value.
+		 *
+		 * @param values The list of default values for the config option
+		 * @return The config option with the default value.
+		 */
+		@SafeVarargs
+		public final ConfigOption<List<T>> defaultValues(T... values) {
+			return new ConfigOption<>(
+				key,
+				clazz,
+				ConfigOption.EMPTY_DESCRIPTION,
+				Arrays.asList(values),
+				true);
+		}
+
+		/**
+		 * Creates a ConfigOption without a default value.
+		 *
+		 * @return The config option without a default value.
+		 */
+		public ConfigOption<List<T>> noDefaultValue() {
+			return new ConfigOption<>(
+				key,
+				clazz,
+				ConfigOption.EMPTY_DESCRIPTION,
+				null,
+				true);
+		}
+	}
+
+
 	// ------------------------------------------------------------------------
 
 	/** Not intended to be instantiated. */
diff --git a/flink-core/src/main/java/org/apache/flink/configuration/Configuration.java b/flink-core/src/main/java/org/apache/flink/configuration/Configuration.java
index 186cf107..79cc7e4c 100644
--- a/flink-core/src/main/java/org/apache/flink/configuration/Configuration.java
+++ b/flink-core/src/main/java/org/apache/flink/configuration/Configuration.java
@@ -25,18 +25,23 @@ import org.apache.flink.core.io.IOReadableWritable;
 import org.apache.flink.core.memory.DataInputView;
 import org.apache.flink.core.memory.DataOutputView;
 import org.apache.flink.types.StringValue;
+import org.apache.flink.util.TimeUtils;
 
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import java.io.IOException;
+import java.time.Duration;
 import java.util.Arrays;
 import java.util.HashMap;
 import java.util.HashSet;
+import java.util.List;
 import java.util.Locale;
 import java.util.Map;
+import java.util.Optional;
 import java.util.Properties;
 import java.util.Set;
+import java.util.stream.Collectors;
 
 import static org.apache.flink.util.Preconditions.checkNotNull;
 
@@ -45,7 +50,7 @@ import static org.apache.flink.util.Preconditions.checkNotNull;
  */
 @Public
 public class Configuration extends ExecutionConfig.GlobalJobParameters
-		implements IOReadableWritable, java.io.Serializable, Cloneable {
+		implements IOReadableWritable, java.io.Serializable, Cloneable, ReadableConfig, WritableConfig {
 
 	private static final long serialVersionUID = 1L;
 
@@ -96,18 +101,21 @@ public class Configuration extends ExecutionConfig.GlobalJobParameters
 	 * @return The value associated with the given key, or the default value, if to entry for the key exists.
 	 */
 	@SuppressWarnings("unchecked")
-	public <T> Class<T> getClass(String key, Class<? extends T> defaultValue, ClassLoader classLoader) throws ClassNotFoundException {
-		Object o = getRawValue(key);
-		if (o == null) {
+	public <T> Class<T> getClass(
+			String key,
+			Class<? extends T> defaultValue,
+			ClassLoader classLoader) throws ClassNotFoundException {
+		Optional<Object> o = getRawValue(key);
+		if (!o.isPresent()) {
 			return (Class<T>) defaultValue;
 		}
 
-		if (o.getClass() == String.class) {
-			return (Class<T>) Class.forName((String) o, true, classLoader);
+		if (o.get().getClass() == String.class) {
+			return (Class<T>) Class.forName((String) o.get(), true, classLoader);
 		}
 
-		LOG.warn("Configuration cannot evaluate value " + o + " as a class name");
-		return (Class<T>) defaultValue;
+		throw new IllegalArgumentException(
+			"Configuration cannot evaluate object of class " + o.get().getClass() + " as a class name");
 	}
 
 	/**
@@ -130,14 +138,13 @@ public class Configuration extends ExecutionConfig.GlobalJobParameters
 	 * @param defaultValue
 	 *        the default value which is returned in case there is no value associated with the given key
 	 * @return the (default) value associated with the given key
+	 * @deprecated use {@link #getString(ConfigOption, String)} or {@link #getOptional(ConfigOption)}
 	 */
+	@Deprecated
 	public String getString(String key, String defaultValue) {
-		Object o = getRawValue(key);
-		if (o == null) {
-			return defaultValue;
-		} else {
-			return o.toString();
-		}
+		return getRawValue(key)
+			.map(this::convertToString)
+			.orElse(defaultValue);
 	}
 
 	/**
@@ -148,8 +155,8 @@ public class Configuration extends ExecutionConfig.GlobalJobParameters
 	 */
 	@PublicEvolving
 	public String getString(ConfigOption<String> configOption) {
-		Object o = getValueOrDefaultFromOption(configOption);
-		return o == null ? null : o.toString();
+		return getOptional(configOption)
+			.orElseGet(configOption::defaultValue);
 	}
 
 	/**
@@ -162,8 +169,8 @@ public class Configuration extends ExecutionConfig.GlobalJobParameters
 	 */
 	@PublicEvolving
 	public String getString(ConfigOption<String> configOption, String overrideDefault) {
-		Object o = getRawValueFromOption(configOption);
-		return o == null ? overrideDefault : o.toString();
+		return getOptional(configOption)
+			.orElse(overrideDefault);
 	}
 
 	/**
@@ -200,14 +207,13 @@ public class Configuration extends ExecutionConfig.GlobalJobParameters
 	 * @param defaultValue
 	 *        the default value which is returned in case there is no value associated with the given key
 	 * @return the (default) value associated with the given key
+	 * @deprecated use {@link #getInteger(ConfigOption, int)} or {@link #getOptional(ConfigOption)}
 	 */
+	@Deprecated
 	public int getInteger(String key, int defaultValue) {
-		Object o = getRawValue(key);
-		if (o == null) {
-			return defaultValue;
-		}
-
-		return convertToInt(o, defaultValue);
+		return getRawValue(key)
+			.map(this::convertToInt)
+			.orElse(defaultValue);
 	}
 
 	/**
@@ -218,8 +224,8 @@ public class Configuration extends ExecutionConfig.GlobalJobParameters
 	 */
 	@PublicEvolving
 	public int getInteger(ConfigOption<Integer> configOption) {
-		Object o = getValueOrDefaultFromOption(configOption);
-		return convertToInt(o, configOption.defaultValue());
+		return getOptional(configOption)
+			.orElseGet(configOption::defaultValue);
 	}
 
 	/**
@@ -233,11 +239,8 @@ public class Configuration extends ExecutionConfig.GlobalJobParameters
 	 */
 	@PublicEvolving
 	public int getInteger(ConfigOption<Integer> configOption, int overrideDefault) {
-		Object o = getRawValueFromOption(configOption);
-		if (o == null) {
-			return overrideDefault;
-		}
-		return convertToInt(o, configOption.defaultValue());
+		return getOptional(configOption)
+			.orElse(overrideDefault);
 	}
 
 	/**
@@ -274,14 +277,13 @@ public class Configuration extends ExecutionConfig.GlobalJobParameters
 	 * @param defaultValue
 	 *        the default value which is returned in case there is no value associated with the given key
 	 * @return the (default) value associated with the given key
+	 * @deprecated use {@link #getLong(ConfigOption, long)} or {@link #getOptional(ConfigOption)}
 	 */
+	@Deprecated
 	public long getLong(String key, long defaultValue) {
-		Object o = getRawValue(key);
-		if (o == null) {
-			return defaultValue;
-		}
-
-		return convertToLong(o, defaultValue);
+		return getRawValue(key)
+			.map(this::convertToLong)
+			.orElse(defaultValue);
 	}
 
 	/**
@@ -292,8 +294,8 @@ public class Configuration extends ExecutionConfig.GlobalJobParameters
 	 */
 	@PublicEvolving
 	public long getLong(ConfigOption<Long> configOption) {
-		Object o = getValueOrDefaultFromOption(configOption);
-		return convertToLong(o, configOption.defaultValue());
+		return getOptional(configOption)
+			.orElseGet(configOption::defaultValue);
 	}
 
 	/**
@@ -307,11 +309,8 @@ public class Configuration extends ExecutionConfig.GlobalJobParameters
 	 */
 	@PublicEvolving
 	public long getLong(ConfigOption<Long> configOption, long overrideDefault) {
-		Object o = getRawValueFromOption(configOption);
-		if (o == null) {
-			return overrideDefault;
-		}
-		return convertToLong(o, configOption.defaultValue());
+		return getOptional(configOption)
+			.orElse(overrideDefault);
 	}
 
 	/**
@@ -348,14 +347,13 @@ public class Configuration extends ExecutionConfig.GlobalJobParameters
 	 * @param defaultValue
 	 *        the default value which is returned in case there is no value associated with the given key
 	 * @return the (default) value associated with the given key
+	 * @deprecated use {@link #getBoolean(ConfigOption, boolean)} or {@link #getOptional(ConfigOption)}
 	 */
+	@Deprecated
 	public boolean getBoolean(String key, boolean defaultValue) {
-		Object o = getRawValue(key);
-		if (o == null) {
-			return defaultValue;
-		}
-
-		return convertToBoolean(o);
+		return getRawValue(key)
+			.map(this::convertToBoolean)
+			.orElse(defaultValue);
 	}
 
 	/**
@@ -366,8 +364,8 @@ public class Configuration extends ExecutionConfig.GlobalJobParameters
 	 */
 	@PublicEvolving
 	public boolean getBoolean(ConfigOption<Boolean> configOption) {
-		Object o = getValueOrDefaultFromOption(configOption);
-		return convertToBoolean(o);
+		return getOptional(configOption)
+			.orElseGet(configOption::defaultValue);
 	}
 
 	/**
@@ -381,11 +379,8 @@ public class Configuration extends ExecutionConfig.GlobalJobParameters
 	 */
 	@PublicEvolving
 	public boolean getBoolean(ConfigOption<Boolean> configOption, boolean overrideDefault) {
-		Object o = getRawValueFromOption(configOption);
-		if (o == null) {
-			return overrideDefault;
-		}
-		return convertToBoolean(o);
+		return getOptional(configOption)
+			.orElse(overrideDefault);
 	}
 
 	/**
@@ -422,14 +417,13 @@ public class Configuration extends ExecutionConfig.GlobalJobParameters
 	 * @param defaultValue
 	 *        the default value which is returned in case there is no value associated with the given key
 	 * @return the (default) value associated with the given key
+	 * @deprecated use {@link #getFloat(ConfigOption, float)} or {@link #getOptional(ConfigOption)}
 	 */
+	@Deprecated
 	public float getFloat(String key, float defaultValue) {
-		Object o = getRawValue(key);
-		if (o == null) {
-			return defaultValue;
-		}
-
-		return convertToFloat(o, defaultValue);
+		return getRawValue(key)
+			.map(this::convertToFloat)
+			.orElse(defaultValue);
 	}
 
 	/**
@@ -440,8 +434,8 @@ public class Configuration extends ExecutionConfig.GlobalJobParameters
 	 */
 	@PublicEvolving
 	public float getFloat(ConfigOption<Float> configOption) {
-		Object o = getValueOrDefaultFromOption(configOption);
-		return convertToFloat(o, configOption.defaultValue());
+		return getOptional(configOption)
+			.orElseGet(configOption::defaultValue);
 	}
 
 	/**
@@ -455,11 +449,8 @@ public class Configuration extends ExecutionConfig.GlobalJobParameters
 	 */
 	@PublicEvolving
 	public float getFloat(ConfigOption<Float> configOption, float overrideDefault) {
-		Object o = getRawValueFromOption(configOption);
-		if (o == null) {
-			return overrideDefault;
-		}
-		return convertToFloat(o, configOption.defaultValue());
+		return getOptional(configOption)
+			.orElse(overrideDefault);
 	}
 
 	/**
@@ -496,14 +487,13 @@ public class Configuration extends ExecutionConfig.GlobalJobParameters
 	 * @param defaultValue
 	 *        the default value which is returned in case there is no value associated with the given key
 	 * @return the (default) value associated with the given key
+	 * @deprecated use {@link #getDouble(ConfigOption, double)} or {@link #getOptional(ConfigOption)}
 	 */
+	@Deprecated
 	public double getDouble(String key, double defaultValue) {
-		Object o = getRawValue(key);
-		if (o == null) {
-			return defaultValue;
-		}
-
-		return convertToDouble(o, defaultValue);
+		return getRawValue(key)
+			.map(this::convertToDouble)
+			.orElse(defaultValue);
 	}
 
 	/**
@@ -514,8 +504,8 @@ public class Configuration extends ExecutionConfig.GlobalJobParameters
 	 */
 	@PublicEvolving
 	public double getDouble(ConfigOption<Double> configOption) {
-		Object o = getValueOrDefaultFromOption(configOption);
-		return convertToDouble(o, configOption.defaultValue());
+		return getOptional(configOption)
+			.orElseGet(configOption::defaultValue);
 	}
 
 	/**
@@ -529,11 +519,8 @@ public class Configuration extends ExecutionConfig.GlobalJobParameters
 	 */
 	@PublicEvolving
 	public double getDouble(ConfigOption<Double> configOption, double overrideDefault) {
-		Object o = getRawValueFromOption(configOption);
-		if (o == null) {
-			return overrideDefault;
-		}
-		return convertToDouble(o, configOption.defaultValue());
+		return getOptional(configOption)
+			.orElse(overrideDefault);
 	}
 
 	/**
@@ -571,20 +558,17 @@ public class Configuration extends ExecutionConfig.GlobalJobParameters
 	 *        The default value which is returned in case there is no value associated with the given key.
 	 * @return the (default) value associated with the given key.
 	 */
-	@SuppressWarnings("EqualsBetweenInconvertibleTypes")
 	public byte[] getBytes(String key, byte[] defaultValue) {
-
-		Object o = getRawValue(key);
-		if (o == null) {
-			return defaultValue;
-		}
-		else if (o.getClass().equals(byte[].class)) {
-			return (byte[]) o;
-		}
-		else {
-			LOG.warn("Configuration cannot evaluate value {} as a byte[] value", o);
-			return defaultValue;
-		}
+		return getRawValue(key).map(o -> {
+				if (o.getClass().equals(byte[].class)) {
+					return (byte[]) o;
+				} else {
+					throw new IllegalArgumentException(String.format(
+						"Configuration cannot evaluate value %s as a byte[] value",
+						o));
+				}
+			}
+		).orElse(defaultValue);
 	}
 
 	/**
@@ -607,8 +591,9 @@ public class Configuration extends ExecutionConfig.GlobalJobParameters
 	 */
 	@PublicEvolving
 	public String getValue(ConfigOption<?> configOption) {
-		Object o = getValueOrDefaultFromOption(configOption);
-		return o == null ? null : o.toString();
+		return Optional.ofNullable(getRawValueFromOption(configOption).orElseGet(configOption::defaultValue))
+			.map(String::valueOf)
+			.orElse(null);
 	}
 
 	/**
@@ -626,15 +611,17 @@ public class Configuration extends ExecutionConfig.GlobalJobParameters
 		checkNotNull(enumClass, "enumClass must not be null");
 		checkNotNull(configOption, "configOption must not be null");
 
-		final String configValue = getString(configOption);
+		Object rawValue = getRawValueFromOption(configOption)
+			.orElseGet(configOption::defaultValue);
 		try {
-			return Enum.valueOf(enumClass, configValue.toUpperCase(Locale.ROOT));
-		} catch (final IllegalArgumentException | NullPointerException e) {
-			final String errorMessage = String.format("Value for config option %s must be one of %s (was %s)",
+			return convertToEnum(rawValue, enumClass);
+		} catch (IllegalArgumentException ex) {
+			final String errorMessage = String.format(
+				"Value for config option %s must be one of %s (was %s)",
 				configOption.key(),
 				Arrays.toString(enumClass.getEnumConstants()),
-				configValue);
-			throw new IllegalArgumentException(errorMessage, e);
+				rawValue);
+			throw new IllegalArgumentException(errorMessage);
 		}
 	}
 
@@ -745,6 +732,36 @@ public class Configuration extends ExecutionConfig.GlobalJobParameters
 		}
 	}
 
+	@Override
+	public <T> T get(ConfigOption<T> option) {
+		return getOptional(option).orElseGet(option::defaultValue);
+	}
+
+	@Override
+	public <T> Optional<T> getOptional(ConfigOption<T> option) {
+		Optional<Object> rawValue = getRawValueFromOption(option);
+		Class clazz = option.getClazz();
+
+		try {
+			if (option.isList()) {
+				return rawValue.map(v -> convertToList(v, clazz));
+			} else {
+				return rawValue.map(v -> convertValue(v, clazz));
+			}
+		} catch (Exception e) {
+			throw new IllegalArgumentException(String.format(
+				"Could not parse value '%s' for key '%s'.",
+				rawValue.map(Object::toString).orElse(""),
+				option.key()), e);
+		}
+	}
+
+	@Override
+	public <T> WritableConfig set(ConfigOption<T> option, T value) {
+		setValueInternal(option.key(), value);
+		return this;
+	}
+
 	// --------------------------------------------------------------------------------------------
 
 	@Override
@@ -799,41 +816,36 @@ public class Configuration extends ExecutionConfig.GlobalJobParameters
 		}
 	}
 
-	private Object getRawValue(String key) {
+	private Optional<Object> getRawValue(String key) {
 		if (key == null) {
 			throw new NullPointerException("Key must not be null.");
 		}
 
 		synchronized (this.confData) {
-			return this.confData.get(key);
+			return Optional.ofNullable(this.confData.get(key));
 		}
 	}
 
-	private Object getRawValueFromOption(ConfigOption<?> configOption) {
+	private Optional<Object> getRawValueFromOption(ConfigOption<?> configOption) {
 		// first try the current key
-		Object o = getRawValue(configOption.key());
+		Optional<Object> o = getRawValue(configOption.key());
 
-		if (o != null) {
+		if (o.isPresent()) {
 			// found a value for the current proper key
 			return o;
 		}
 		else if (configOption.hasFallbackKeys()) {
 			// try the deprecated keys
 			for (FallbackKey fallbackKey : configOption.fallbackKeys()) {
-				Object oo = getRawValue(fallbackKey.getKey());
-				if (oo != null) {
+				Optional<Object> oo = getRawValue(fallbackKey.getKey());
+				if (oo.isPresent()) {
 					loggingFallback(fallbackKey, configOption);
 					return oo;
 				}
 			}
 		}
 
-		return null;
-	}
-
-	private Object getValueOrDefaultFromOption(ConfigOption<?> configOption) {
-		Object o = getRawValueFromOption(configOption);
-		return o != null ? o : configOption.defaultValue();
+		return Optional.empty();
 	}
 
 	private void loggingFallback(FallbackKey fallbackKey, ConfigOption<?> configOption) {
@@ -850,99 +862,173 @@ public class Configuration extends ExecutionConfig.GlobalJobParameters
 	//  Type conversion
 	// --------------------------------------------------------------------------------------------
 
-	private int convertToInt(Object o, int defaultValue) {
+	@SuppressWarnings("unchecked")
+	private <T> T convertValue(Object rawValue, Class clazz) {
+		if (Integer.class.equals(clazz)) {
+			return (T) convertToInt(rawValue);
+		} else if (Long.class.equals(clazz)) {
+			return (T) convertToLong(rawValue);
+		} else if (Boolean.class.equals(clazz)) {
+			return (T) convertToBoolean(rawValue);
+		} else if (Float.class.equals(clazz)) {
+			return (T) convertToFloat(rawValue);
+		} else if (Double.class.equals(clazz)) {
+			return (T) convertToDouble(rawValue);
+		} else if (String.class.equals(clazz)) {
+			return (T) convertToString(rawValue);
+		} else if (clazz.isEnum()) {
+			return (T) convertToEnum(rawValue, clazz);
+		} else if (clazz == Duration.class) {
+			return (T) convertToDuration(rawValue);
+		} else if (clazz == MemorySize.class) {
+			return (T) convertToMemorySize(rawValue);
+		} else if (clazz == Map.class) {
+			return (T) convertToProperties(rawValue);
+		}
+
+		throw new IllegalArgumentException("Unsupported type: " + clazz);
+	}
+
+	@SuppressWarnings("unchecked")
+	private <T> T convertToList(Object rawValue, Class atomicClass) {
+		if (rawValue instanceof List) {
+			return (T) rawValue;
+		} else {
+			return (T) StructuredOptionsSplitter.splitEscaped(rawValue.toString(), ';').stream()
+				.map(s -> convertValue(s, atomicClass))
+				.collect(Collectors.toList());
+		}
+	}
+
+	@SuppressWarnings("unchecked")
+	private Map<String, String> convertToProperties(Object o) {
+		if (o instanceof Map) {
+			return (Map<String, String>) o;
+		} else {
+			List<String> listOfRawProperties = StructuredOptionsSplitter.splitEscaped(o.toString(), ',');
+			return listOfRawProperties.stream()
+				.map(s -> StructuredOptionsSplitter.splitEscaped(s, ':'))
+				.map(pair -> {
+					if (pair.size() != 2) {
+						throw new IllegalArgumentException("Could not parse pair in the map " + pair);
+					}
+
+					return pair;
+				})
+				.collect(Collectors.toMap(
+					a -> a.get(0),
+					a -> a.get(1)
+				));
+		}
+	}
+
+	@SuppressWarnings("unchecked")
+	private <E extends Enum<E>> E convertToEnum(Object o, Class<E> clazz) {
+		if (o.getClass().equals(clazz)) {
+			return (E) o;
+		}
+
+		return Enum.valueOf(clazz, o.toString().toUpperCase(Locale.ROOT));
+	}
+
+	private Duration convertToDuration(Object o) {
+		if (o.getClass() == Duration.class) {
+			return (Duration) o;
+		}
+
+		return TimeUtils.parseDuration(o.toString());
+	}
+
+	private MemorySize convertToMemorySize(Object o) {
+		if (o.getClass() == MemorySize.class) {
+			return (MemorySize) o;
+		}
+
+		return MemorySize.parse(o.toString());
+	}
+
+	private String convertToString(Object o) {
+		if (o.getClass() == String.class) {
+			return (String) o;
+		} else if (o.getClass() == Duration.class) {
+			Duration duration = (Duration) o;
+			return String.format("%d ns", duration.toNanos());
+		}
+
+		return o.toString();
+	}
+
+	private Integer convertToInt(Object o) {
 		if (o.getClass() == Integer.class) {
 			return (Integer) o;
-		}
-		else if (o.getClass() == Long.class) {
+		} else if (o.getClass() == Long.class) {
 			long value = (Long) o;
 			if (value <= Integer.MAX_VALUE && value >= Integer.MIN_VALUE) {
 				return (int) value;
 			} else {
-				LOG.warn("Configuration value {} overflows/underflows the integer type.", value);
-				return defaultValue;
-			}
-		}
-		else {
-			try {
-				return Integer.parseInt(o.toString());
-			}
-			catch (NumberFormatException e) {
-				LOG.warn("Configuration cannot evaluate value {} as an integer number", o);
-				return defaultValue;
+				throw new IllegalArgumentException(String.format(
+					"Configuration value %s overflows/underflows the integer type.",
+					value));
 			}
 		}
+
+		return Integer.parseInt(o.toString());
 	}
 
-	private long convertToLong(Object o, long defaultValue) {
+	private Long convertToLong(Object o) {
 		if (o.getClass() == Long.class) {
 			return (Long) o;
-		}
-		else if (o.getClass() == Integer.class) {
+		} else if (o.getClass() == Integer.class) {
 			return ((Integer) o).longValue();
 		}
-		else {
-			try {
-				return Long.parseLong(o.toString());
-			}
-			catch (NumberFormatException e) {
-				LOG.warn("Configuration cannot evaluate value " + o + " as a long integer number");
-				return defaultValue;
-			}
-		}
+
+		return Long.parseLong(o.toString());
 	}
 
-	private boolean convertToBoolean(Object o) {
+	private Boolean convertToBoolean(Object o) {
 		if (o.getClass() == Boolean.class) {
 			return (Boolean) o;
 		}
-		else {
-			return Boolean.parseBoolean(o.toString());
+
+		switch (o.toString().toUpperCase()) {
+			case "TRUE":
+				return true;
+			case "FALSE":
+				return false;
+			default:
+				throw new IllegalArgumentException(String.format(
+					"Unrecognized option for boolean: %s. Expected either true or false(case insensitive)",
+					o));
 		}
 	}
 
-	private float convertToFloat(Object o, float defaultValue) {
+	private Float convertToFloat(Object o) {
 		if (o.getClass() == Float.class) {
 			return (Float) o;
-		}
-		else if (o.getClass() == Double.class) {
+		} else if (o.getClass() == Double.class) {
 			double value = ((Double) o);
 			if (value == 0.0
 					|| (value >= Float.MIN_VALUE && value <= Float.MAX_VALUE)
 					|| (value >= -Float.MAX_VALUE && value <= -Float.MIN_VALUE)) {
 				return (float) value;
 			} else {
-				LOG.warn("Configuration value {} overflows/underflows the float type.", value);
-				return defaultValue;
-			}
-		}
-		else {
-			try {
-				return Float.parseFloat(o.toString());
-			}
-			catch (NumberFormatException e) {
-				LOG.warn("Configuration cannot evaluate value {} as a float value", o);
-				return defaultValue;
+				throw new IllegalArgumentException(String.format(
+					"Configuration value %s overflows/underflows the float type.",
+					value));
 			}
 		}
+
+		return Float.parseFloat(o.toString());
 	}
 
-	private double convertToDouble(Object o, double defaultValue) {
+	private Double convertToDouble(Object o) {
 		if (o.getClass() == Double.class) {
 			return (Double) o;
-		}
-		else if (o.getClass() == Float.class) {
+		} else if (o.getClass() == Float.class) {
 			return ((Float) o).doubleValue();
 		}
-		else {
-			try {
-				return Double.parseDouble(o.toString());
-			}
-			catch (NumberFormatException e) {
-				LOG.warn("Configuration cannot evaluate value {} as a double value", o);
-				return defaultValue;
-			}
-		}
+
+		return Double.parseDouble(o.toString());
 	}
 
 	// --------------------------------------------------------------------------------------------
@@ -984,7 +1070,8 @@ public class Configuration extends ExecutionConfig.GlobalJobParameters
 						value = bytes;
 						break;
 					default:
-						throw new IOException("Unrecognized type: " + type);
+						throw new IOException(String.format("Unrecognized type: %s. This method is deprecated and" +
+							" might not work for all supported types.", type));
 				}
 
 				this.confData.put(key, value);
@@ -1035,7 +1122,8 @@ public class Configuration extends ExecutionConfig.GlobalJobParameters
 					out.writeBoolean((Boolean) val);
 				}
 				else {
-					throw new IllegalArgumentException("Unrecognized type");
+					throw new IllegalArgumentException("Unrecognized type. This method is deprecated and might not work" +
+						" for all supported types.");
 				}
 			}
 		}
@@ -1052,7 +1140,6 @@ public class Configuration extends ExecutionConfig.GlobalJobParameters
 		return hash;
 	}
 
-	@SuppressWarnings("EqualsBetweenInconvertibleTypes")
 	@Override
 	public boolean equals(Object obj) {
 		if (this == obj) {
diff --git a/flink-core/src/main/java/org/apache/flink/configuration/DelegatingConfiguration.java b/flink-core/src/main/java/org/apache/flink/configuration/DelegatingConfiguration.java
index b0249a05..450325b9 100644
--- a/flink-core/src/main/java/org/apache/flink/configuration/DelegatingConfiguration.java
+++ b/flink-core/src/main/java/org/apache/flink/configuration/DelegatingConfiguration.java
@@ -29,6 +29,7 @@ import java.util.HashMap;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
+import java.util.Optional;
 import java.util.Properties;
 import java.util.Set;
 
@@ -332,6 +333,21 @@ public final class DelegatingConfiguration extends Configuration {
 		return backingConfig.contains(prefixOption(configOption, prefix));
 	}
 
+	@Override
+	public <T> T get(ConfigOption<T> option) {
+		return backingConfig.get(prefixOption(option, prefix));
+	}
+
+	@Override
+	public <T> Optional<T> getOptional(ConfigOption<T> option) {
+		return backingConfig.getOptional(prefixOption(option, prefix));
+	}
+
+	@Override
+	public <T> WritableConfig set(ConfigOption<T> option, T value) {
+		return backingConfig.set(prefixOption(option, prefix), value);
+	}
+
 	// --------------------------------------------------------------------------------------------
 
 	@Override
@@ -379,9 +395,12 @@ public final class DelegatingConfiguration extends Configuration {
 		}
 
 		FallbackKey[] deprecated = deprecatedKeys.toArray(new FallbackKey[0]);
-		return new ConfigOption<>(key,
+		return new ConfigOption<T>(
+			key,
+			option.getClazz(),
 			option.description(),
 			option.defaultValue(),
+			option.isList(),
 			deprecated);
 	}
 }
diff --git a/flink-core/src/main/java/org/apache/flink/configuration/ReadableConfig.java b/flink-core/src/main/java/org/apache/flink/configuration/ReadableConfig.java
new file mode 100644
index 00000000..9b4c3419
--- /dev/null
+++ b/flink-core/src/main/java/org/apache/flink/configuration/ReadableConfig.java
@@ -0,0 +1,53 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.configuration;
+
+import org.apache.flink.annotation.PublicEvolving;
+
+import java.util.Optional;
+
+/**
+ * Read access to a configuration object. Allows reading values described with meta information
+ * included in {@link ConfigOption}.
+ */
+@PublicEvolving
+public interface ReadableConfig {
+
+	/**
+	 * Reads a value using the metada included in {@link ConfigOption}. Returns the
+	 * {@link ConfigOption#defaultValue()} if value key not present in the configuration.
+	 *
+	 * @param option metadata of the option to read
+	 * @param <T> type of the value to read
+	 * @return read value or {@link ConfigOption#defaultValue()} if not found
+	 * @see #getOptional(ConfigOption)
+	 */
+	<T> T get(ConfigOption<T> option);
+
+	/**
+	 * Reads a value using the metada included in {@link ConfigOption}. In contrast to
+	 * {@link #get(ConfigOption)} returns {@link Optional#empty()} if value not present.
+	 *
+	 * @param option metadata of the option to read
+	 * @param <T> type of the value to read
+	 * @return read value or {@link Optional#empty()} if not found
+	 * @see #get(ConfigOption)
+	 */
+	<T> Optional<T> getOptional(ConfigOption<T> option);
+}
diff --git a/flink-core/src/main/java/org/apache/flink/configuration/StructuredOptionsSplitter.java b/flink-core/src/main/java/org/apache/flink/configuration/StructuredOptionsSplitter.java
new file mode 100644
index 00000000..44ae82ad
--- /dev/null
+++ b/flink-core/src/main/java/org/apache/flink/configuration/StructuredOptionsSplitter.java
@@ -0,0 +1,176 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.configuration;
+
+import org.apache.flink.annotation.Internal;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * Helper class for splitting a string on a given delimiter with quoting logic.
+ */
+@Internal
+class StructuredOptionsSplitter {
+
+	/**
+	 * Splits the given string on the given delimiter. It supports quoting parts of the string with
+	 * either single (') or double quotes ("). Quotes can be escaped by doubling the quotes.
+	 *
+	 * <p>Examples:
+	 * <ul>
+	 *     <li>'A;B';C => [A;B], [C]</li>
+	 *     <li>"AB'D";B;C => [AB'D], [B], [C]</li>
+	 *     <li>"AB'""D;B";C => [AB'\"D;B], [C]</li>
+	 * </ul>
+	 *
+	 * <p>For more examples check the tests.
+	 * @param string a string to split
+	 * @param delimiter delimiter to split on
+	 * @return a list of splits
+	 */
+	static List<String> splitEscaped(String string, char delimiter) {
+		List<Token> tokens = tokenize(checkNotNull(string), delimiter);
+		return processTokens(tokens);
+	}
+
+	private static List<String> processTokens(List<Token> tokens) {
+		final List<String> splits = new ArrayList<>();
+		for (int i = 0; i < tokens.size(); i++) {
+			Token token = tokens.get(i);
+			switch (token.getTokenType()) {
+				case DOUBLE_QUOTED:
+				case SINGLE_QUOTED:
+					if (i + 1 < tokens.size() && tokens.get(i + 1).getTokenType() != TokenType.DELIMITER) {
+						int illegalPosition = tokens.get(i + 1).getPosition() - 1;
+						throw new IllegalArgumentException(
+							"Could not split string. Illegal quoting at position: " + illegalPosition);
+					}
+					splits.add(token.getString());
+					break;
+				case UNQUOTED:
+					splits.add(token.getString());
+					break;
+				case DELIMITER:
+					if (i + 1 < tokens.size() && tokens.get(i + 1).getTokenType() == TokenType.DELIMITER) {
+						splits.add("");
+					}
+					break;
+			}
+		}
+
+		return splits;
+	}
+
+	private static List<Token> tokenize(String string, char delimiter) {
+		final List<Token> tokens = new ArrayList<>();
+		final StringBuilder builder = new StringBuilder();
+		for (int cursor = 0; cursor < string.length(); ) {
+			final char c = string.charAt(cursor);
+
+			int nextChar = cursor + 1;
+			if (c == '\'') {
+				nextChar = consumeInQuotes(string, '\'', cursor, builder);
+				tokens.add(new Token(TokenType.SINGLE_QUOTED, builder.toString(), cursor));
+			} else if (c == '"') {
+				nextChar = consumeInQuotes(string, '"', cursor, builder);
+				tokens.add(new Token(TokenType.DOUBLE_QUOTED, builder.toString(), cursor));
+			} else if (c == delimiter) {
+				tokens.add(new Token(TokenType.DELIMITER, String.valueOf(c), cursor));
+			} else if (!Character.isWhitespace(c)) {
+				nextChar = consumeUnquoted(string, delimiter, cursor, builder);
+				tokens.add(new Token(TokenType.UNQUOTED, builder.toString().trim(), cursor));
+			}
+			builder.setLength(0);
+			cursor = nextChar;
+		}
+
+		return tokens;
+	}
+
+	private static int consumeInQuotes(String string, char quote, int cursor, StringBuilder builder) {
+		for (int i = cursor + 1; i < string.length(); i++) {
+			char c = string.charAt(i);
+			if (c == quote) {
+				if (i + 1 < string.length() && string.charAt(i + 1) == quote) {
+					builder.append(c);
+					i += 1;
+				} else {
+					return i + 1;
+				}
+			} else {
+				builder.append(c);
+			}
+		}
+
+		throw new IllegalArgumentException("Could not split string. Quoting was not closed properly.");
+	}
+
+	private static int consumeUnquoted(String string, char delimiter, int cursor, StringBuilder builder) {
+		int i;
+		for (i = cursor; i < string.length(); i++) {
+			char c = string.charAt(i);
+			if (c == delimiter) {
+				return i;
+			} else if (c == '\'' || c == '"') {
+				throw new IllegalArgumentException("Could not split string. Illegal quoting at position: " + i);
+			}
+
+			builder.append(c);
+		}
+
+		return i;
+	}
+
+	private enum TokenType {
+		DOUBLE_QUOTED,
+		SINGLE_QUOTED,
+		UNQUOTED,
+		DELIMITER
+	}
+
+	private static class Token {
+		private final TokenType tokenType;
+		private final String string;
+		private final int position;
+
+		private Token(TokenType tokenType, String string, int position) {
+			this.tokenType = tokenType;
+			this.string = string;
+			this.position = position;
+		}
+
+		public TokenType getTokenType() {
+			return tokenType;
+		}
+
+		public String getString() {
+			return string;
+		}
+
+		public int getPosition() {
+			return position;
+		}
+	}
+
+	private StructuredOptionsSplitter() {
+	}
+}
diff --git a/flink-core/src/main/java/org/apache/flink/configuration/WritableConfig.java b/flink-core/src/main/java/org/apache/flink/configuration/WritableConfig.java
new file mode 100644
index 00000000..594f5b77
--- /dev/null
+++ b/flink-core/src/main/java/org/apache/flink/configuration/WritableConfig.java
@@ -0,0 +1,40 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.configuration;
+
+import org.apache.flink.annotation.PublicEvolving;
+
+/**
+ * Write access to a configuration object. Allows storing values described with meta information
+ * included in {@link ConfigOption}.
+ */
+@PublicEvolving
+public interface WritableConfig {
+
+	/**
+	 * Stores a given value using the metadata included in the {@link ConfigOption}.
+	 * The value should be readable back through {@link ReadableConfig}.
+	 *
+	 * @param option metadata information
+	 * @param value value to be stored
+	 * @param <T> type of the value to be stored
+	 * @return instance of this configuration for fluent API
+	 */
+	<T> WritableConfig set(ConfigOption<T> option, T value);
+}
